server/channel_mob.go

```diff
@@ -39,21 +50,52 @@ func (server ChannelServer) mobControl(conn mnet.Client, reader mpacket.Reader)
 
 	moveBytes := movement.GenerateMovementBytes(moveData)
 
-	inst.LifePool().MobAcknowledge(mobSpawnID, plr, moveID, skillPossible, byte(action), skillData, moveData, finalData, moveBytes)
+	skillDelay := int16(skillData >> 16)
+	skillLevel := byte(skillData >> 8)
+	skillID := byte(skillData)
+
+	if actualAction >= 21 && actualAction <= 25 {
+		mob.PerformSkill(skillDelay, skillLevel, skillID)
+	} else if actualAction > 12 && actualAction < 20 {
+		attackID := byte(actualAction - 12)
+
+		// check mob can use attack
+		if level, valid := mob.Skills[attackID]; valid {
+			levels, err := nx.GetMobSkill(attackID)
+
+			if err != nil {
+				return
+			}
+
+			if int(level) < len(levels) {
+				skill := levels[level]
+				mob.SetMP(mob.MP() - skill.MpCon)
+				if mob.MP() < 0 {
+					mob.SetMP(0)
+				}
+			}
+
+		}
 
-	// skillDelay := int16(skillData >> 16)
-	// skillID := byte(skillData)
-	// skillLevel := byte(skillData >> 8)
+		mob.SetLastAttackTime(time.Now().Unix())
+		mob.PerformAttack(attackID)
+	}
+
+	// Calculate the next action
+	mob.CanUseSkill = skillPossible
+
+	if !mob.CanUseSkill || (mob.StatBuff()&skills.MobStat.SealSkill > 0) {
+		// there are more reasons as to why a mob cannot use a skill
+		skillID = 0
+	} else {
+		skillID, skillLevel = mob.GetNextSkill()
+	}
+
+	inst.LifePool().MobAcknowledge(mobSpawnID, plr, moveID, skillPossible, byte(action), skillData, moveData, finalData, moveBytes)
 
-	// if actualAction >= 21 && actualAction <= 25 {
-	// 	mob.PerformSkill(skillDelay, skillLevel, skillID)
-	// } else if actualAction > 12 && actualAction < 20 {
-	// 	mob.PerformAttack(byte(actualAction - 12))
-	// }
+	// This will update the next skill but currently in a loop
+	//mob.AcknowledgeController(moveID, finalData, skillPossible, skillID, skillLevel)
```

Comment:
What do you mean by "it is in a loop"? Does the mob do the same skills straight away after finishing the previous one?

Reply:
Sorry let me explain more. 

Once a mob begins attacking it will find a random skill for that mob to use. The skill will then be applied to the mob and it will use that skill. For the moment I have the skill cooldown set at 10 seconds so during those 10 seconds the function on line 91 returns 0,0 until the cooldown is reached. Once cooldown has been reached then it will get a new skill.

During the 10 seconds while the function is returning 0,0 the mob uses the same skill that it was given at the beginning. It does not revert itself back to "normal" attacks during this time, and will just constantly use that same skill over and over until a new one is given. Ideally, what I want to do is use the skill, and then after the skill is used set the mob back to normal attacks.

Reply:
That seems odd as I think that is controlled by `skillPossible` and `skillID` variables. When I had it previously implemented the mob was definitely respecting the skills usage cooldown. I suspect the false is more important than the skill level, id and mp from looking at the packet: 

https://github.com/Hucaru/Valhalla/blob/master/server/field/lifepool/mob/packet.go#L23-L28

I wouldn't be surprised if the client reads the bool and if it's false stops reading the packet. Other things to try would be setting the mp to zero and see if that forces it to stop. You could also try inverting the boolean value as it might not technically be a boolean (0 false, 1 true) in the client.

> function on line 91 returns 0,0 until the cooldown is reached

Have you verified this is actually happening with a print statement or otherwise?

Reply:
I think I know where the bug is. skillPossible doesn't observe the timer. So everytime the controller tells the server I want a skill (line 24) it will always hit the else block. Even though you return 0,0 the skillPossible flag is true.

Reply:
Yes, I have verified it is return the proper variables. I believe I found the issue. I was previously sending the values to 2 different controllers. I updated it to create a new skilldata variable and update with the proper data. It seems to be working correctly now.

Reply:
I edited my previous comment with an issue with the function in general.